app/gui/view/graph-editor/src/builtin/visualization/native/text_visualization.rs

```diff
@@ -0,0 +1,469 @@
+/// messages sent to the visualisation as well as the caching performance. A larger value will
+/// result in fewer, smaller messages, but the visualisation might have to load more data that is
+/// not needed, as it will be cropped. For example, a value of 100, would mean that the
+/// visualisation would request 100 characters per chunk, even if it can only show 10 characters at
+/// once in the available viewport.
+const CHARS_PER_CHUNK: usize = 20;
+/// Extra chunks to load around the visible grid to ensure smooth scrolling. Extra chunks are
+/// loaded in each direction around the visible grid. So a value of 5 with a base grid of 20x10 will
+/// load 25x15 grid.
+const CACHE_PADDING: u32 = 15;
+const PADDING_TEXT: f32 = 5.0;
+
+
+
+// =============================
+// === Grid Coordinate Types ===
+// =============================
+
+type GridPosition = Vector2<i32>;
+type GridSize = Vector2<i32>;
+type GridVector = Vector2<i32>;
+
+/// Position and size of a sub-grid within a karger grid.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
+pub struct GridWindow {
+    position: GridPosition,
+    size:     GridSize,
+}
+
+use text_provider::BackendTextProvider;
+use text_provider::TextProvider;
+
+
+
+// =============
+// === Model ===
+// =============
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct Model<T> {
+    app:                   Application,
+    size:                  Rc<Cell<Vector2>>,
+    root:                  display::object::Instance,
+    // Note that we are using a simple `GridView` and our own scrollbar, instead of the
+    // `scrollable::GridView` to avoid adding `ScrollAreas` to the scene, as the clipping they
+    // provide though the `mask::View` is not free in terms of performance (they add a draw call
+    // cost) and we don't need it here because we need to clip DOM elements anyway.
+    text_grid:             GridView<grid_view_entry::Entry>,
+    dom_entry_root:        web::HtmlDivElement,
+    clipping_div:          DomSymbol,
+    scroll_bar_horizontal: Scrollbar,
+    scroll_bar_vertical:   Scrollbar,
+    text_provider:         Rc<RefCell<Option<T>>>,
+}
+
+impl<T: 'static> Model<T> {
+    /// Constructor.
+    fn new(app: Application) -> Self {
+        let scene = &app.display.default_scene;
+        let root = display::object::Instance::new();
+        let clipping_div = web::document.create_div_or_panic();
+        let clipping_div = DomSymbol::new(&clipping_div);
+        let dom_entry_root = web::document.create_div_or_panic();
+        let size = default();
+        let text_provider = default();
+
+        clipping_div.set_style_or_warn("overflow", "hidden");
+        dom_entry_root.set_style_or_warn("position", "absolute");
+        scene.dom.layers.front.manage(&clipping_div);
+        root.add_child(&clipping_div);
+        clipping_div.append_or_warn(&dom_entry_root);
```

Comment:
This is super complex. Without proper explanation it is not understandable. Please refactor to a well documented function